open-machine-learning-jupyter-book/ml-fundamentals/Build a Web App to use a Mchine Learning model.md

```diff
@@ -0,0 +1,352 @@
+
+```{code-cell}
+import pandas as pd
+import numpy as np
+    
+ufos = pd.read_csv('../assets/data/ufos.csv')
+ufos.head()
+```
+
+Convert the ufos data to a small dataframe with fresh titles. Check the unique values in the `Country` field.
+
+```{code-cell}
+ufos = pd.DataFrame({'Seconds': ufos['duration (seconds)'], 'Country': ufos['country'],'Latitude': ufos['latitude'],'Longitude': ufos['longitude']})
+
+ufos.Country.unique()
+```
+
+Now, you can reduce the amount of data we need to deal with by dropping any null values and only importing sightings between 1-60 seconds:
+
+```{code-cell}
+ufos.dropna(inplace=True)
+
+ufos = ufos[(ufos['Seconds'] >= 1) & (ufos['Seconds'] <= 60)]
+
+ufos.info()
+```
+
+Import Scikit-learn's `LabelEncoder` library to convert the text values for countries to a number:
+
+
+```{code-cell}
+from sklearn.preprocessing import LabelEncoder
+
+ufos['Country'] = LabelEncoder().fit_transform(ufos['Country'])
+
+ufos.head()
+```
+
+```{note}
+LabelEncoder encodes data alphabetically
+```
+
+## Exercise - build your model
+
+Now you can get ready to train a model by dividing the data into the training and testing group.
+
+ Select the three features you want to train on as your X vector, and the y vector will be the `Country`. You want to be able to input `Seconds`, `Latitude` and `Longitude` and get a country id to return.
+
+```{code-cell}
+from sklearn.model_selection import train_test_split
+
+Selected_features = ['Seconds','Latitude','Longitude']
+
+X = ufos[Selected_features].values
+y = ufos['Country'].values
+
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
+```
+
+ Train your model using logistic regression:
+
+```{code-cell}
+from sklearn.metrics import accuracy_score, classification_report
+from sklearn.linear_model import LogisticRegression
+model = LogisticRegression(solver="sag",max_iter=10000)
+model.fit(X_train, y_train)
+predictions = model.predict(X_test)
+
+print(classification_report(y_test, predictions))
+print('Predicted labels: ', predictions)
+print('Accuracy: ', accuracy_score(y_test, predictions))
+```
+
+The accuracy isn't bad **(around 95%)**, unsurprisingly, as `Country` and `Latitude/Longitude` correlate.
+
+The model you created isn't very revolutionary as you should be able to infer a `Country` from its `Latitude` and `Longitude`, but it's a good exercise to try to train from raw data that you cleaned, exported, and then use this model in a web app.
+
+## Exercise - 'pickle' your model
+
+Now, it's time to _pickle_ your model! You can do that in a few lines of code. Once it's _pickled_, load your pickled model and test it against a sample data array containing values for seconds, latitude and longitude,
+
+```{code-cell}
+import pickle
+model_filename = 'ufo-model.pkl'
+pickle.dump(model, open(model_filename,'wb'))
+
+model = pickle.load(open('ufo-model.pkl','rb'))
```

Comment:
The pickle file should not be in this folder. Instead, move it into `assets/pickle/` and then update the path. Similar to how you handle the csv